### PR TITLE
Fix statistic of json type

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Fix what
Misestimate the cost of plan node than contains the JSON type. E.g. the following SCAN costs `0` though it has 6001215 rows of data.

```sql
MySQL sb_json> explain logical select count(*) from lineorder_flat_json;
+---------------------------------------------------------------------------------------+
| Explain String                                                                        |
+---------------------------------------------------------------------------------------+
| - Output => [4:count]                                                                 |
|     - AGGREGATE(GLOBAL) []                                                            |
|             Estimates: {row: 1, cpu: 0.00, memory: 8.00, network: 0.00}               |
|             4:count := count()                                                        |
|         - EXCHANGE(GATHER)                                                            |
|                 Estimates: {row: 6001215, cpu: 0.00, memory: 0.00, network: 0.00}     |
|             - SCAN [lineorder_flat_json] => [3:j]                                     |
|                     Estimates: {row: 6001215, cpu: 0.00, memory: 0.00, network: 0.00} |
|                     partitionRatio: 7/7, tabletRatio: 336/336                         |
+---------------------------------------------------------------------------------------+
9 rows in set
```

## Modification
1. Fix typeSize estimation of JSON
2. Use `char_length` function to calculate the row-size of JSON
3. Modify the Statistic, make the least row-size to 1, avoid huge cost estimate error

## Problem Summary(Required) ：

Statistic for JSON type is inaccurate, which make `select count(*) from json_table` generates one stage aggregation plan, which is in-efficient.

### CODE REVIEW

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-2.2

File: fe/fe-core/src/main/java/com/starrocks/catalog/PrimitiveType.java [hunk 1]
```diff
@@ -377,6 +377,11 @@ public int getTypeSize() {
                 // 1MB
                 typeSize = 1024 * 1024;
                 break;
+            case JSON:
+                typeSize = 1024;
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/catalog/PrimitiveType.java:381]: 1kb is too small for json type?

Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/catalog/PrimitiveType.java:381]: As SSB flat dataset, convert the table row to JSON costs about 800+ bytes per JSON field.

**Comment by kangkaisen** [comment]
> Statistic for JSON type is inaccurate, which make select count(*) from json_table generates one stage aggregation plan, which is in-efficient.

please  add  UT cover  it.

File: fe/fe-core/src/main/java/com/starrocks/catalog/PrimitiveType.java (hunk 1 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/catalog/PrimitiveType.java:381]: I don't konw, but th user's data isn't like SSB only int and short-string, maybe we need remeber to adjust the size

Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/catalog/PrimitiveType.java:381]: Actually in normal case the statistic executor don't need to use this typeSize to estimate the row-size, it use `char_length` function to touch the the real data.

**Review by HangyuanLiu — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 9 / 12 (75.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/PrimitiveType.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/catalog/Table.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/Statistics.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/statistic/StatisticExecutor.java|1|1|100.00%|

**Review by Seaven — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/Statistics.java [hunk 2]
```diff
@@ -32,17 +32,23 @@ public double getOutputRowCount() {
...
     public double getOutputSize(ColumnRefSet outputColumns) {
         double totalSize = 0;
+        boolean nonEmpty = false;
         for (Map.Entry<ColumnRefOperator, ColumnStatistic> entry : columnStatistics.entrySet()) {
             if (outputColumns.contains(entry.getKey().getId())) {
                 totalSize += entry.getValue().getAverageRowSize();
+                nonEmpty = true;
             }
         }
+        if (nonEmpty) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/Statistics.java:42]: outputColumns not in columnStatistics? May be the outputColumns and column statistics were not calculated correctly

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/Statistics.java [hunk 3]
```diff
@@ -32,17 +32,23 @@ public double getOutputRowCount() {
...
+            totalSize = Math.max(totalSize, 1.0);
+        }
         return totalSize * outputRowCount;
     }
 
     public double getComputeSize() {
-        return this.columnStatistics.values().stream().map(ColumnStatistic::getAverageRowSize).
-                reduce(0.0, Double::sum) * outputRowCount;
+        // Make it at least 1 byte, otherwise the cost model would propagate estimate error
+        return Math.max(1.0, this.columnStatistics.values().stream().map(ColumnStatistic::getAverageRowSize).
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/Statistics.java:50]: Why column statistics average rowsize could less than 1 byte？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/Statistics.java (hunk 2 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/Statistics.java:42]: Just for the sake of conservation

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/Statistics.java (hunk 3 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/Statistics.java:50]: A column's row-size could be zero, when all rows are null.

In this case, the statistic of JSON is calculated wrongly, and it's 0.

**Review by Youngwb — APPROVED** [review]

**Pull request merged**